derivation/src/main/scala-2/caliban/derivation/DerivationMacros.scala

```diff
@@ -0,0 +1,423 @@
+        q"""
+            _root_.caliban.introspection.adt.__EnumValue(
+              name = ${subtypeInfo.name},
+              description = ${subtypeInfo.description},
+              isDeprecated = ${subtypeInfo.isDeprecated},
+              deprecationReason = ${subtypeInfo.deprecationReason}
+            )
+         """
+      }
+
+    q"""
+          ${refs.makeEnum}(
+            Some(${info.name}),
+            ${info.description},
+            List(..$enumValues),
+            None
+          )
+    """
+  }
+
+  private def deriveStep(
+    parent: Type,
+    info: GraphQLInfo,
+    env: Type,
+    resolveValue: TermName,
+    outputs: List[TermSymbol]
+  ): Tree = {
+    val fields =
+      outputs.map(o => DeriveMember(o, env).deriveStepWithName(parent, resolveValue))
+
+    q"""
+      ${refs.ObjectStep}.apply[$env](
+        ${info.name},
+        Map(..$fields)
+      )
+    """
+  }
+
+  private def mergeEnvironments(members: List[TermSymbol]): Type = {
+    val envTypes = members.map(m => DeriveMember(m, typeRefs.Any).envType).toSet - typeRefs.Any
+    if (envTypes.nonEmpty) {
+      envTypes.reduce((a, b) => tq"$a with $b".tpe)
+    } else {
+      typeRefs.Any
+    }
+  }
+
+  private def deriveProductSchema(tpe: Type, requestedEnv: Type): Tree = {
+    val inputs  = inputFields(tpe)
+    val outputs = outputFields(tpe)
+
+    val envType = mergeEnvironments(inputs ++ outputs)
+
+    if (!(requestedEnv <:< requestedEnv)) {
+      c.warning(
+        c.enclosingPosition,
+        s"deriveSchema was called with environment ${requestedEnv} but there are members in the derived type is using ${envType}!"
+      )
+    }
+
+    val resolveValue = TermName(c.freshName())
+
+    val info    = GraphQLInfo(tpe.typeSymbol)
+    val toType  =
+      q"""
+        override def toType(isInput: Boolean = false, isSubscription: Boolean = false): ${typeRefs.__Type} =
+          if (isInput) ${deriveInputObject(info, requestedEnv, inputs)}
+          else ${deriveOutputObject(info, requestedEnv, outputs)}
+      """
+    val resolve =
+      q"""
+         override def resolve($resolveValue: $tpe): ${Step(requestedEnv)} =
+           ${deriveStep(tpe, info, requestedEnv, resolveValue, outputs)}
+      """
+
+    val result = q"""
+      new ${Schema(requestedEnv, tpe)} {
+        $toType
+        $resolve
+      }
+    """
+    result
+  }
+
+  def deriveSumSchema(tpe: Type, requestedEnv: Type): Tree = {
+    // NOTE: Sealed trait cannot have inputs
+    val outputs    = outputFields(tpe)
+    val subclasses = knownSubclassesOf(tpe.typeSymbol.asClass)
+
+    val subclassInOut =
+      subclasses.map { subclass =>
+        val inputs  = inputFields(subclass.typeSignature)
+        val outputs = outputFields(subclass.typeSignature)
+
+        (subclass, (inputs, outputs))
+      }.toMap
+
+    val isEnum      = outputs.isEmpty && subclassInOut.forall { case (_, (in, out)) => in.isEmpty && out.isEmpty }
```

Comment:
Recently was added the ability to use union instead of enum using the `GQLUnion` (and the `fixEmptyUnionObject` dirty trick).